EngineLayer/PeptideSpectralMatch.cs

```diff
@@ -393,5 +402,122 @@ private PeptideSpectralMatch(PeptideSpectralMatch psm, List<(int Notch, PeptideW
             PeptidesToMatchingFragments = psm.PeptidesToMatchingFragments;
             SpectralAngle = psm.SpectralAngle;
         }
+
+        /// <summary>
+        /// Determine the Fragment Coverage the PSM
+        /// Assigns fragment coverage indices for the PSM and the protein based on Matched Ion Fragments
+        /// </summary>
+        public void AddFragmentCoveragePSMs()
+        {
+            if (this.PeptideLength == null)
+            {
+                return;
+            }
+
+            if (this.BaseSequence == null)
+            {
+                return;
+            }
+
+            if (this.MatchedFragmentIons == null)
+            {
+                return;
+            }
+            var peptideLength = this.PeptideLength.Value;
+            var peptideLengthMinus1 = peptideLength - 1;
+            List<int> bIonPositions = new();
+            List<int> yIonPositions = new();
+
+            // only look at b and y ions
+            var ionsbAndy = this.MatchedFragmentIons.Where(p => p.NeutralTheoreticalProduct.ProductType == Proteomics.Fragmentation.ProductType.b ||p.NeutralTheoreticalProduct.ProductType == Proteomics.Fragmentation.ProductType.y).ToList();
+            foreach (var ion in ionsbAndy)
+            {
+                switch (ion.NeutralTheoreticalProduct.ProductType)
+                {
+                    // put each ion position into a list, either b or y
+                    case Proteomics.Fragmentation.ProductType.b:
+                        bIonPositions.Add(ion.NeutralTheoreticalProduct.FragmentNumber);
+                        break;
+                    case Proteomics.Fragmentation.ProductType.y:
+                        yIonPositions.Add(ion.NeutralTheoreticalProduct.FragmentNumber);
+                        break;
+                }
+            }
```

Comment:
Do we want to account for internal ions as well?

Reply:
that point might be worth a discussion.

Reply:
I think internal ion fragment coverage should be a separate pull request. Including internal ions in fragment coverage is a broader conversation in the field as a whole, so we should make a determination on that at a later time.

Reply:
My thought is that for the highest confidence of single amino acid coverage we would only want to use terminal b and y fragments. But I would defer to other opinions if people feel strongly.